active/deployment-principal-function.md

```diff
@@ -0,0 +1,121 @@
+---
+REP Number: <Fill me in with a four-digit number matching the pull request number; Update AFTER PR is approved and BEFORE is merged.>
+Author: levimatheri (Levi Muriuki)
+Start Date: 2024-09-09
+Feature Status: Private Preview
+Bicep Issue Number(s): [#645](https://github.com/Azure/bicep/issues/645), [#4959](https://github.com/Azure/bicep/issues/4959), [#9969](https://github.com/Azure/bicep/discussions/9969)
+---
+
+# Title - New deployment principal function
+
+## Summary
+
+Introduces a new function to get the principal that is submitting the current deployment. This can be a user or an application (service principal)
+
+## Terms and definitions
+
+## Motivation
+
+Customers have expressed that having a function that can output properties about the current deployment principal would be helpful for [use-cases](#use-cases-expressed-by-customers) that require certain fields such as `objectId` (`principalId`). The current workarounds to achieve this are not desirable as these workarounds include running a prerequisite `az` command step to obtain the current principal's fields and then passing them into the Bicep template as parameters, e.g.
+
+```sh
+az login
+let principalId = az rest --method GET --uri https://graph.microsoft.com/v1.0/me | from json | get id
+az deployment group create -g mygroup -f my.bicep -p $"principalId=($principalId)"
+```
+
+```bicep
+param principalId string
+...
+
+resource roleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+  ...
+  properties: {
+    ...
+    principalId: principalId
+    principalType: 'User'
+  }
+}
+```
+
+### Use-cases expressed by customers
+- Assigning the current principal to RBAC roles to various resources within the same template
+    - [API definition](https://learn.microsoft.com/en-us/azure/templates/microsoft.authorization/roleassignments?pivots=deployment-language-bicep)
+- Adding current user to a Entra ID group created using Bicep extensibility 
+    - requires member's `objectId`
+- Setting the default admin of a SQL server
+    - [API definition](https://learn.microsoft.com/en-us/azure/templates/microsoft.sql/servers/administrators?pivots=deployment-language-bicep)
+- Assigning Key Vault access policies
+    - [API definition](https://learn.microsoft.com/en-us/azure/templates/microsoft.keyvault/vaults/accesspolicies?pivots=deployment-language-bicep#accesspolicyentry)
+
+## Detailed design
+- We plan to expose a minimal set of properties to satisfy the above use-cases, namely:
+    - `objectId` - represents the unique identifier of the principal within a tenant
+    - `type` - i.e. `User`, `Service Principal`
```

Comment:
[nit] To be consistent with other Azure types (which are often based on C# enums), I'd go with `ServicePrincipal` (without the space).

Reply:
On this one, I'd also make sure that it can be supported feasibly by the service - it'll need to be something available on incoming requests from ARM.